pkg/services/searchV2/filter.go

```diff
@@ -74,28 +75,35 @@ func (q *PermissionFilter) canAccess(kind entityKind, id string) bool {
 	//
 	switch kind {
 	case entityKindFolder:
-		if id == "" {
+		if id == folder.GeneralFolderUID {
 			q.logAccessDecision(true, kind, id, "generalFolder")
 			return true
 		}
-		fallthrough
+		decision := q.filter(entityKindFolder, id, "")
+		q.logAccessDecision(decision, kind, id, "resourceFilter")
+		return decision
 	case entityKindDashboard:
-		decision := q.filter(id)
+		decision := q.filter(entityKindDashboard, id, location)
 		q.logAccessDecision(decision, kind, id, "resourceFilter")
 		return decision
 	case entityKindPanel:
 		matches := panelIdFieldRegex.FindStringSubmatch(id)
-
 		submatchCount := len(matches)
 		if submatchCount != panelIdFieldRegexExpectedSubmatchCount {
 			q.logAccessDecision(false, kind, id, "invalidPanelIdFieldRegexSubmatchCount", "submatchCount", submatchCount, "expectedSubmatchCount", panelIdFieldRegexExpectedSubmatchCount)
 			return false
 		}
-
 		dashboardUid := matches[panelIdFieldDashboardUidSubmatchIndex]
-		decision := q.filter(dashboardUid)
 
-		q.logAccessDecision(decision, kind, id, "resourceFilter", "dashboardUid", dashboardUid, "panelId", matches[panelIdFieldPanelIdSubmatchIndex])
+		// Location is <folder_uid>/<dashboard_uid>
+		if len(location) <= len(dashboardUid)+1 {
```

Comment:
```suggestion
		// Location is <folder_uid>/<dashboard_uid>
		if !strings.HasSuffix(location, "/"+dashboardUid) {
```

would this be a stronger way of checking this constraint? also, I haven't checked, but isn't an empty string a valid `folderUID` representing the root folder?

Reply:
From what I can see, general folder is represented with UID "general". I don't know exactly where this translation happens, so chances are it doesn't apply to all cases. I'll dig through the code a bit to make sure there are no exceptions.

Reply:
Right, I think this is where `general` is set for folder UID for the root folder: https://github.com/grafana/grafana/blob/main/pkg/services/searchV2/index.go#L763-L765

We should probably use a constant there.

I wonder if this bit should also be changed by setting UID to `general`?
https://github.com/grafana/grafana/blob/main/pkg/services/searchV2/index.go#L907-L919

Reply:
We also use UID `general` for folders with ID 0 in access control evaluations: [here](https://github.com/grafana/grafana/blob/main/pkg/services/dashboards/accesscontrol.go#L137-L139) and [here](https://github.com/grafana/grafana/blob/main/pkg/services/dashboards/accesscontrol.go#L72-L74).